### DISCUSSION TITLE
Is there a way to authenticate the user in a Node script?

### DISCUSSION THREAD

**Discussion by csalmeida** [supabase/supabase#15729] [Questions] [answered]
Hi all,

I am trying to set up a script to seed the database programmatically. I am aware that there's a `seed.sql` file method but in this case, we're doing it with a Node script.

I have setup a `seed.ts` file that runs statements like ` supabase.auth.admin.createUser()`  or `supabase.from('posts').insert(options)`.

This works for the most part, but our setup runs triggers to complete records requiring the user related to that data to be signed in. Or other checks might fail that use `auth.id()`:

```sql
CREATE POLICY "Insert post" ON posts
    FOR INSERT
    with CHECK (
        auth.uid() is not NULL
    );
```

Is there any way to overcome this? I am using the service key instead of the anon key and have tried `supabase.auth.setAuth(jwt)` but it doesn't seem to no longer exist and the error I get using `signInWithPassword` is the following:

```ts
   // No storage option exists to persist the session, which may result in unexpected behavior when using auth.
   // If you want to set persistSession to true, please provide a storage option or you may set persistSession to false to disable this warning.
  const { data: session } = await supabase.auth.signInWithPassword({
    email,
    password,
  })
```

**Comment by GaryAustin1** [comment]
You should not do `auth.uid() is not Null`.  You should use `TO authenticated` for the policy it is much faster.

Either way though, you can generate a jwt with role claim `authenticated` and a sub claim with the UUID (or a made up UUID If it is not associated with a real user in auth.users table).  Then you createClient with a the global header option and set a Authorization header to 'Bearer jwt'.   Then you don't need to sign in.
Your persist session error is because you are on a server without local storage.  You can set persistSession:false in the createClient to get rid of that if desired.

**Reply by csalmeida** [reply]
> You should not do `auth.uid() is not Null`. You should use `TO authenticated` for the policy it is much faster.

I didn't know that, thanks for letting me know.

> Either way though, you can generate a jwt with role claim `authenticated` and a sub claim with the UUID (or a made up UUID If it is not associated with a real user in auth.users table). Then you createClient with a the global header option and set a Authorization header to 'Bearer jwt'. Then you don't need to sign in.

Thank you, this seems like something I could try. Is there any documentation on how to do this?

> Your persist session error is because you are on a server without local storage. You can set persistSession:false in the createClient to get rid of that if desired.

Nice one, I found out later that it can be disabled. 

Thanks a lot for the help, I'll update this post if I manage to get it work. 🙏

**Reply by csalmeida** [reply]
I think a found [an example in this other discussion](https://github.com/orgs/supabase/discussions/1094#discussioncomment-587587). Will give it a go when I have a minute.

```javascript
const supabase = createClient(supabaseUrl, supabaseKey, {
  headers: {
    apiKey: anonKey,
    Authorisation: `Bearer ${jwt}`
  }
});
```

I need to find where I get the JWT from this seems to be related:

- https://github.com/supabase/gotrue-js/issues/34
- https://github.com/supabase/gotrue-js/pull/80

**Comment by csalmeida** [comment, marked as answer]
Here's how I solved it, I created a function that signs the user in and use that JWT to generate another client to run queries that required the `auth.id()` parameter:

```ts
export async function supabaseJWTClient(userCredentials: {
  email: string
  password: string
}) {
  // Used to sign user in:
  const supabase = createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.SUPABASE_SERVICE_KEY!,
    {
      auth: {
        persistSession: false,
      },
    }
  )

  // Sign user in and extract JWT token:
  const { data: authenticated } = await supabase.auth.signInWithPassword({
    email: userCredentials.email,
    password: userCredentials.password,
  })

  const jwt = authenticated.session?.access_token
  
  // Pass token to a new client and run the queries that rely on a user signed in:
  return createClient(process.env.NEXT_PUBLIC_SUPABASE_URL!, jwt!, {
    auth: {
      persistSession: false, // Hides session storage warning from showing in terminal output.
    },
  })
}
```

Then I instantiate the client that makes use of the JWT to run queries that run triggers with the user ID:

```ts
const credentials: {
  email: 'sup@supabase.com',
  password: 'password'
}

const supabase = await supabaseJWTClient(credentials)
  
const { data: post, error: postError } = await supabase
  .from('posts')
  .insert(options)
  .select()
  .single()
```

Using this method the triggers related with inserts to the `posts` table that use `auth.id()` can run without throwing an errors like the following:

```js
  postError {
    code: '23502',
    details: 'Failing row contains (84a0a062-fc85-472b-a6a7-e3af7434f9da, null, 1, 2023-07-29 15:50:44.87067+00, 05c4b590-adf6-4c00-ba05-264454ef9c3f).',
    hint: null,
    message: 'null value in column "user_id" of relation "post_authors" violates not-null constraint'
  }
```
